### PR TITLE
GH-45: Add RECEIVED... Headers

### LINKED ISSUE #45: Consider Adding KafkaHeaders.RECEIVED_* Headers
Separate inbound headers from outbound to prevent propagation by default - allows the elimination of the `enableHeaderRouting` property in the s-i-k `KPMH`.

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves #45

Avoid propagation of received headers when sending `o.s.messaging.Message`s.

Also add `Message<?>` methods to the template.

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java [hunk 1]
```diff
@@ -211,9 +233,21 @@ public RecordMetadata syncSend(String topic, int partition, K key, V data)
...
-	public void flush() {
-		this.producer.flush();
+	private ProducerRecord<K, V> messageToProducerRecord(Message<?> message) {
+		MessageHeaders headers = message.getHeaders();
+		Object topic = headers.get(KafkaHeaders.TOPIC);
+		Object partition = headers.get(KafkaHeaders.PARTITION_ID);
+		Object key = headers.get(KafkaHeaders.MESSAGE_KEY);
+		Object payload = message.getPayload();
+		Assert.isTrue(topic == null || topic instanceof String, "topic header must be a String");
+		Assert.isTrue(partition == null || partition instanceof Integer, "partition header must be an Integer");
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java]: Can't we use an overloaded version of `MessageHeaders.get()` with the `Class` to cast instead?

**Comment by artembilan** [comment]
Here is a polishing how I assume them: https://github.com/artembilan/spring-kafka/commit/213cfcc3d9afaa3718e4102050af3409b4e3face

**Comment by garyrussell** [comment]
Doh; thanks; pushed - also fixed those `send` methods.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/753d9bdb89ca5dcf22ee0a18b8da10303ed13c73

**Pull request closed without merging**